geode-redis/src/distributedTest/java/org/apache/geode/redis/internal/executor/hash/HKeysDUnitTest.java

```diff
@@ -0,0 +1,159 @@
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.geode.redis.internal.executor.hash;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import redis.clients.jedis.Jedis;
+
+import org.apache.geode.redis.ConcurrentLoopingThreads;
+import org.apache.geode.test.awaitility.GeodeAwaitility;
+import org.apache.geode.test.dunit.rules.MemberVM;
+import org.apache.geode.test.dunit.rules.RedisClusterStartupRule;
+
+public class HKeysDUnitTest {
+
+  @ClassRule
+  public static RedisClusterStartupRule clusterStartUp = new RedisClusterStartupRule(4);
+
+  private static final String LOCAL_HOST = "127.0.0.1";
+  private static final int HASH_SIZE = 1000;
+  private static final int NUM_ITERATIONS = 10000;
+  private static final int JEDIS_TIMEOUT =
+      Math.toIntExact(GeodeAwaitility.getTimeout().toMillis());
+  private static Jedis jedis1;
+  private static Jedis jedis2;
+
+  private static Properties locatorProperties;
+
+  private static MemberVM locator;
+  private static MemberVM server1;
+  private static MemberVM server2;
+
+  private static int redisServerPort1;
+  private static int redisServerPort2;
+
+  @BeforeClass
+  public static void classSetup() {
+    locatorProperties = new Properties();
+
+    locator = clusterStartUp.startLocatorVM(0, locatorProperties);
+    server1 = clusterStartUp.startRedisVM(1, locator.getPort());
+    server2 = clusterStartUp.startRedisVM(2, locator.getPort());
+
+    redisServerPort1 = clusterStartUp.getRedisPort(1);
+    redisServerPort2 = clusterStartUp.getRedisPort(2);
+
+    jedis1 = new Jedis(LOCAL_HOST, redisServerPort1, JEDIS_TIMEOUT);
+    jedis2 = new Jedis(LOCAL_HOST, redisServerPort2, JEDIS_TIMEOUT);
+  }
+
+  @Before
+  public void testSetup() {
+    jedis1.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis1.disconnect();
+    jedis2.disconnect();
+
+    server1.stop();
+    server2.stop();
+    locator.stop();
+  }
+
+  @Test
+  public void testConcurrentHKeys_whileAddingValues() {
+    String key = "key";
+
+    Map<String, String> testMap = makeHashMap(HASH_SIZE, "field-", "value-");
+    Set<String> expectedFields = makeSet(HASH_SIZE, "field-");
+
+    jedis1.hset(key, testMap);
+
+    new ConcurrentLoopingThreads(NUM_ITERATIONS,
+        (i) -> {
+          int newIndex = HASH_SIZE + i;
+          jedis1.hset(key, "field-" + newIndex, "value-" + newIndex);
+          expectedFields.add("field-" + newIndex);
+        },
+        (i) -> assertThat(jedis2.hkeys(key)).containsExactlyInAnyOrderElementsOf(expectedFields));
```

Comment:
This and the other uses of `ConcurrentLoopingThreads` are missing the `.run()` call to actually execute them. Unfortunately you'll see that they cause a `ConcurrentModificationException` on the `expectedFields` variable. Even so, there would be a race between lines 102 and 103 and the assertion on 105 that does the check on the hash. It's always hard to know what to assert on in this kind of a case. You could probably just adjust the assertion to be `containsAll()`.